tests/Temporalio.Tests/Worker/WorkflowWorkerTests.cs

```diff
@@ -3012,6 +3012,62 @@ public async Task ExecuteWorkflowAsync_TaskEvents_AreRecordedProperly()
             workerOptions);
     }
 
+    public class DuplicateActivities
+    {
+        private readonly string returnValue;
+
+        public DuplicateActivities(string returnValue) => this.returnValue = returnValue;
+
+        [Activity]
+        public string DoThing() => returnValue;
+    }
+
+    [Workflow]
+    public class DuplicateActivityWorkflow
+    {
+        [WorkflowRun]
+        public async Task<string[]> RunAsync(string taskQueue1, string taskQueue2)
+        {
+            return new[]
+            {
+                await Workflow.ExecuteActivityAsync(
+                    (DuplicateActivities act) => act.DoThing(),
+                    new()
+                    {
+                        TaskQueue = taskQueue1,
+                        ScheduleToCloseTimeout = TimeSpan.FromHours(1),
+                    }),
+                await Workflow.ExecuteActivityAsync(
+                    (DuplicateActivities act) => act.DoThing(),
+                    new()
+                    {
+                        TaskQueue = taskQueue2,
+                        ScheduleToCloseTimeout = TimeSpan.FromHours(1),
+                    }),
+            };
+        }
+    }
+
+    [Fact]
+    public async Task ExecuteWorkflowAsync_DuplicateActivity_DoesNotCacheInstance()
+    {
+        await ExecuteWorkerAsync<DuplicateActivityWorkflow>(
+            async worker1 =>
+            {
+                await ExecuteWorkerAsync<DuplicateActivityWorkflow>(
+                    async worker2 =>
+                    {
+                        var ret = await Env.Client.ExecuteWorkflowAsync(
+                            (DuplicateActivityWorkflow wf) =>
+                                wf.RunAsync(worker1.Options.TaskQueue!, worker2.Options.TaskQueue!),
+                            new(id: $"workflow-{Guid.NewGuid()}", taskQueue: worker1.Options.TaskQueue!));
+                        Assert.Equal(new[] { "instance1", "instance2" }, ret);
+                    },
+                    new TemporalWorkerOptions().AddAllActivities(new DuplicateActivities("instance2")));
+            },
+            new TemporalWorkerOptions().AddAllActivities(new DuplicateActivities("instance1")));
+    }
+
```

Comment:
Explaining this test to myself:

1.We define a workflow that returns a result containing the results of two activity calls.

2. The two activities are calls to the same activity method but on distinct instances (and using distinct task queues)

3. The utility `ExecuteWorkerAsync` allows us to start a worker and, while it's running, execute an arbitrary function that has access to the worker instance. We use it twice:

4. The first usage starts the first worker and uses the function to make the second nested call to `ExecuteWorkerAsync`

5. The second usage starts a second worker and uses the function to execute our workflow and check that the result shows that distinct activity methods were invoked.

Reply:
So we're using the lambda passed to `ExecuteWorkerAsync` to essentially play the role of customer application (non-Worker) code, since it starts a workflow. Therefore it's a bit artificial that the function has access to an actual instance of the worker (this confused me initially). But it looks like the pass-a-function-that-receives-a-worker-instance is a deeper part of non-test part of the SDK design.

Reply:
Bingo. Before I removed the cache, this test would fail because the result of the workflow would be `["instance1","instance1"]` because no matter which activity instance you gave to a worker, it used the first cached one based on method which was bad. Now it returns `["instance1","instance2"]` as expected.

Reply:
> So we're using the lambda passed to ExecuteWorkerAsync to essentially play the role of customer application (non-Worker) code, since it starts a workflow.

Yes

> Therefore it's a bit artificial that the function has access to an actual instance of the worker (this confused me initially).

Mostly. It's just so I can get it to create that random task queue which I can use

> But it looks like the pass-a-function-that-receives-a-worker-instance is a deeper part of the non-test-code SDK design.

Basically you have a worker that can run for some amount of time. In non-test code we support a worker running forever, until cancellation token is triggered, and/or until the completion of an async function. In test code, I made `ExecuteWorkerAsync` which is just a shortcut for the latter that also adds a workflow and makes a random task queue and does a couple of other things and gives the worker as the parameter just in case the test function wants to do something with it (most just use the task queue).